prov/gni/src/gnix_cq.c

```diff
@@ -39,12 +39,293 @@
+	entry = NULL;
+err:
+	return entry;
+}
+
+
+static void fill_cq_entry(void *cq_entry, void *op_context, uint64_t flags,
+			  size_t len, void *buf, uint64_t data, uint64_t tag)
+{
+	struct fi_cq_entry *entry = cq_entry;
 
+	entry->op_context = op_context;
+}
+
+static void fill_cq_msg(void *cq_entry, void *op_context, uint64_t flags,
+			size_t len, void *buf, uint64_t data, uint64_t tag)
+{
+	struct fi_cq_msg_entry *entry = cq_entry;
+
+	entry->op_context = op_context;
+	entry->flags = flags;
+	entry->len = len;
+}
+
+static void fill_cq_data(void *cq_entry, void *op_context, uint64_t flags,
+			 size_t len, void *buf, uint64_t data, uint64_t tag)
+{
+	struct fi_cq_data_entry *entry = cq_entry;
+
+	entry->op_context = op_context;
+	entry->flags = flags;
+	entry->len = len;
+	entry->buf = buf;
+	entry->data = data;
+}
+
+static void fill_cq_tagged(void *cq_entry, void *op_context, uint64_t flags,
+			   size_t len, void *buf, uint64_t data, uint64_t tag)
+{
+	struct fi_cq_tagged_entry *entry = cq_entry;
+
+	entry->op_context = op_context;
+	entry->flags = flags;
+	entry->buf = buf;
+	entry->data = data;
+	entry->tag = tag;
+}
+
+static inline void gnix_cq_enqueue(struct gnix_fid_cq *cq,
+				   struct gnix_cq_entry *event)
+{
+	if (!cq || !event)
+		return;
+
+	fastlock_acquire(&cq->lock);
+
+	/*
+	 * TODO: Handle wait signal support.
+	 */
+	slist_insert_tail(&event->item, &cq->ev_queue);
+
+	fastlock_release(&cq->lock);
+}
+
+static ssize_t gnix_cq_dequeue(struct gnix_fid_cq *cq, void *buf, size_t count,
+			       fi_addr_t *src_addr)
+{
+	struct gnix_cq_entry *event;
+	ssize_t read_count = 0;
+
+	if (slist_empty(&cq->ev_queue))
+		return -FI_EAGAIN;
+
+	fastlock_acquire(&cq->lock);
+
+	while(count--) {
+		event = container_of(slist_remove_head(&cq->ev_queue),
+				     struct gnix_cq_entry,
+				     item);
+
+		memcpy(buf, event->the_entry, cq->entry_size);
+		slist_insert_tail(&event->item, &cq->ev_free);
+
+		buf += cq->entry_size;
+		src_addr++;
+
+		read_count++;
+	}
+
+	fastlock_release(&cq->lock);
+
+	return read_count;
+}
+
+static ssize_t gnix_cq_add_event(struct gnix_fid_cq *cq, void *op_context,
+				 uint64_t flags, size_t len, void *buf,
+				 uint64_t data, uint64_t tag)
+{
+	struct gnix_cq_entry *event;
+
+	fastlock_acquire(&cq->lock);
+
+	if (!slist_empty(&cq->ev_free)) {
+		event = container_of(slist_remove_head(&cq->ev_free),
+				     struct gnix_cq_entry, item);
+	} else {
+		event = alloc_cq_entry(cq->entry_size);
+	}
+
+	fastlock_release(&cq->lock);
+
+	if (!event) {
+		GNIX_ERR(FI_LOG_CQ, "error creating cq_entry\n");
+		return -FI_ENOMEM;
+	}
+
+	
+	fill_function[cq->attr.format](event, op_context, flags, len, buf,
+				       data, tag);
+
+	gnix_cq_enqueue(cq, event);
+
+	return FI_SUCCESS;
+}
+
+static ssize_t gnix_cq_add_error(struct gnix_fid_cq *cq, void *op_context,
+				 uint64_t flags, size_t len, void *buf,
+				 uint64_t data, uint64_t tag, size_t olen,
+				 int err, int prov_errno, void *err_data)
```

Comment:
Since this will hopefully be a rare thing and not need to be high performance, might as well provide as many fields as we can.  We will also want to include the prov_errno.